### PR TITLE
GAWB-3052: rejection handling

### PR DESCRIPTION
**Pull request by rushtong**
Rejections are slightly different between spray/akka-http. In the current repo, we only care about one error case where the requested content type is incorrect and respond with a nice error. 

This PR:
* Adds the older version of handling the custom case to a custom handler.
* Re-adds the one case where tests were custom handling a rejection.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/agora/server/webservice/ApiService.scala [hunk 1]
```diff
@@ -54,6 +54,15 @@ object ApiService extends LazyLogging with SprayJsonSupport with DefaultJsonProt
         complete(InternalServerError, AgoraException(e.getMessage, e.getCause, InternalServerError))
     }
   }
+
+  val rejectionHandler: RejectionHandler =
+    RejectionHandler.newBuilder().handle {
+      case MalformedRequestContentRejection(message, cause) => complete(BadRequest, message)
```
Comment (rushtong) [inline src/main/scala/org/broadinstitute/dsde/agora/server/webservice/ApiService.scala]: This was only used in `ApiServiceSpec` but tests relied on this being handled in this way which is why I added it back to all routes.

File: src/main/scala/org/broadinstitute/dsde/agora/server/webservice/ApiService.scala [hunk 2]
```diff
@@ -54,6 +54,15 @@ object ApiService extends LazyLogging with SprayJsonSupport with DefaultJsonProt
...
+    }.handle {
+      case UnacceptedResponseContentTypeRejection(supported) => complete(NotAcceptable,
+        "Resource representation is only available with these Content-Types:\n" + supported.map(_.format).mkString("\n"))
```
Comment (rushtong) [inline src/main/scala/org/broadinstitute/dsde/agora/server/webservice/ApiService.scala]: This is how `spray` handled this case in `RejectionHandler`, but akka-http handles it slightly differently. Carrying the original approach over to maintain a friendly error response.

**Review by davidangb — COMMENTED** [review]
for DRY and conciseness ... how about create
```
def handleExceptionsAndRejections = handleExceptions(exceptionHandler) & handleRejections(rejectionHandler)
```
inside the `ApiService` object. Then, everywhere else you added those directives, you can just use 
```
ApiService.handleExceptionsAndRejections { ... }
```
instead.

**Comment by rushtong** [comment]
@davidangb - thanks, great idea. Will implement.

**Review by aednichols — COMMENTED** [review]
Seems reasonable, I will circle back once the changes are in

File: src/main/scala/org/broadinstitute/dsde/agora/server/webservice/ApiService.scala [hunk 3]
```diff
@@ -32,14 +32,18 @@ object ApiService extends LazyLogging with SprayJsonSupport with DefaultJsonProt
...
   val exceptionHandler: ExceptionHandler = {
     ExceptionHandler {
       case withErrorReport: WorkbenchExceptionWithErrorReport =>
         complete(withErrorReport.errorReport.statusCode.getOrElse(InternalServerError), withErrorReport.errorReport)
       case workbenchException: WorkbenchException =>
         val report = ErrorReport(Option(workbenchException.getMessage).getOrElse(""), Some(InternalServerError), Seq(), Seq(), Some(workbenchException.getClass))
         complete(InternalServerError, report)
-      case e: IllegalArgumentException => complete(BadRequest, e)
+      case e: IllegalArgumentException =>
+        logger.info("ApiService: IllegalArgumentException")
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/agora/server/webservice/ApiService.scala]: do you need this logging?

**Review by davidangb — APPROVED** [review]
:+1: lgtm!

**Review by aednichols — APPROVED** [review]
👍 Tested every combination of (text, json) x (WDL, plain-WDL), works as expected.

File: src/main/scala/org/broadinstitute/dsde/agora/server/webservice/ApiService.scala (hunk 3 above)
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/agora/server/webservice/ApiService.scala]: oops, no, that's debugging.

**Review by aednichols — APPROVED** [review]
👍 Cool, very unobtrusive design with this latest change

**Review by davidangb — APPROVED** [review]
still 👍

**Pull request merged**